### PR TITLE
Update Eclipse Mosquitto to 1.5.3.

### PR DESCRIPTION
**Pull request by ralight**
This is an update for the Eclipse Mosquitto image. I'm the upstream lead and am taking on the direct maintainership now.

Changes:
* Multiple architecture support
* The new Dockerfile builds from source instead of using an Alpine package
* The downloads have sha256 and gpg checks.

It is a fairly substantial rewrite and my Docker experience isn't as broad as I'd like, so I'd be grateful for a keen review.

### CODE REVIEW

**Comment by yosifkit** [comment]
<details>
<summary>Diff:</summary>

```diff
diff --git a/_bashbrew-arches b/_bashbrew-arches
index d69cbdb..86f7445 100644
--- a/_bashbrew-arches
+++ b/_bashbrew-arches
@@ -1 +1,7 @@
 eclipse-mosquitto:latest @ amd64
+eclipse-mosquitto:latest @ arm32v6
+eclipse-mosquitto:latest @ arm32v7
+eclipse-mosquitto:latest @ arm64v8
+eclipse-mosquitto:latest @ i386
+eclipse-mosquitto:latest @ ppc64le
+eclipse-mosquitto:latest @ s390x
diff --git a/_bashbrew-list b/_bashbrew-list
index 82c7fd6..55dfd82 100644
--- a/_bashbrew-list
+++ b/_bashbrew-list
@@ -1,2 +1,3 @@
-eclipse-mosquitto:1.4.12
+eclipse-mosquitto:1.5
+eclipse-mosquitto:1.5.3
 eclipse-mosquitto:latest
diff --git a/eclipse-mosquitto_latest/Dockerfile b/eclipse-mosquitto_latest/Dockerfile
index be20a14..5106830 100644
--- a/eclipse-mosquitto_latest/Dockerfile
+++ b/eclipse-mosquitto_latest/Dockerfile
@@ -1,13 +1,69 @@
-FROM alpine:3.6
-MAINTAINER David Audet <david.audet@ca.com>
+FROM alpine:3.8
 
-LABEL Description="Eclipse Mosquitto MQTT Broker"
+LABEL maintainer="Roger Light <roger@atchoo.org>" \
+	 description="Eclipse Mosquitto MQTT Broker"
 
-RUN apk --no-cache add mosquitto=1.4.12-r0 && \
+ENV VERSION=1.5.3 \
+	 DOWNLOAD_SHA256=3081a998d303a883b1cd064009beabc88aa9159e26f5258a4ae6007160491d10 \
+	 GPG_KEYS=A0D6EEA1DCAE49A635A3B2F0779B22DFB3E717B7
+
+RUN set -x && \
+	 apk --no-cache add --virtual build-deps \
+      build-base \
+      curl \
+	  gnupg \
+      libressl-dev \
+      libwebsockets-dev \
+      util-linux-dev && \
+    curl -fSL https://mosquitto.org/files/source/mosquitto-${VERSION}.tar.gz -o /tmp/mosq.tar.gz && \
+	echo "$DOWNLOAD_SHA256  /tmp/mosq.tar.gz" | sha256sum -c - && \
+    curl -fSL https://mosquitto.org/files/source/mosquitto-${VERSION}.tar.gz.asc -o /tmp/mosq.tar.gz.asc && \
+	export GNUPGHOME="$(mktemp -d)" && \
+	found=''; \
+	for server in \
+		ha.pool.sks-keyservers.net \
+		hkp://keyserver.ubuntu.com:80 \
+		hkp://p80.pool.sks-keyservers.net:80 \
+		pgp.mit.edu \
+	; do \
+		echo "Fetching GPG key $GPG_KEYS from $server"; \
+		gpg --keyserver "$server" --keyserver-options timeout=10 --recv-keys "$GPG_KEYS" && found=yes && break; \
+	done; \
+	test -z "$found" && echo >&2 "error: failed to fetch GPG key $GPG_KEYS" && exit 1; \
+	gpg --batch --verify /tmp/mosq.tar.gz.asc /tmp/mosq.tar.gz && \
+	gpgconf --kill all && \
+	rm -rf "$GNUPGHOME" /tmp/mosq.tar.gz.asc && \
+	mkdir -p /build && \
+	tar --strip=1 -xf /tmp/mosq.tar.gz -C /build && \
+	rm /tmp/mosq.tar.gz && \
+    make -C /build -j "$(nproc)" \
+      WITH_ADNS=no \
+	  WITH_DOCS=no \
+      WITH_MEMORY_TRACKING=no \
+	  WITH_SHARED_LIBRARIES=no \
+      WITH_SRV=no \
+	  WITH_STRIP=yes \
+      WITH_TLS_PSK=no \
+      WITH_WEBSOCKETS=yes \
+      prefix=/usr \
+      binary && \
+	addgroup -S mosquitto 2>/dev/null && \
+    adduser -S -D -H -h /var/empty -s /sbin/nologin -G mosquitto -g mosquitto mosquitto 2>/dev/null && \
     mkdir -p /mosquitto/config /mosquitto/data /mosquitto/log && \
-    cp /etc/mosquitto/mosquitto.conf /mosquitto/config && \
-    chown -R mosquitto:mosquitto /mosquitto
+	install -d /usr/sbin/ && \
+	install -s -m755 /build/src/mosquitto /usr/sbin/mosquitto && \
+	install -m644 /build/mosquitto.conf /mosquitto/config/mosquitto.conf && \
+    sed -i -e 's/#log_dest stderr/log_dest syslog/' /mosquitto/config/mosquitto.conf && \
+    chown -R mosquitto:mosquitto /mosquitto && \
+	apk del build-deps && \
+	apk --no-cache add \
+      libuuid \
+      libwebsockets && \
+	rm -rf /build /etc/apk /lib/apk
+
+VOLUME ["/mosquitto/config", "/mosquitto/data", "/mosquitto/log"]
 
+# Set up the entry point script and default command
 COPY docker-entrypoint.sh /
 ENTRYPOINT ["/docker-entrypoint.sh"]
 CMD ["/usr/sbin/mosquitto", "-c", "/mosquitto/config/mosquitto.conf"]
diff --git a/eclipse-mosquitto_latest/docker-entrypoint.sh b/eclipse-mosquitto_latest/docker-entrypoint.sh
index 1a9fc8d..b381ac5 100755
--- a/eclipse-mosquitto_latest/docker-entrypoint.sh
+++ b/eclipse-mosquitto_latest/docker-entrypoint.sh
@@ -1,5 +1,4 @@
 #!/bin/ash
-
 set -e
-exec "$@"
 
+exec "$@"
```

</details>

**Comment by yosifkit** [comment]
Just a minor suggestion on the volumes: https://github.com/docker-library/official-images/pull/2437#issuecomment-266578827.  So I would suggest that `/mosquitto/config` might not be a great target for a volume.

The rest of the Dockerization looks fine to me.

**Comment by ralight** [comment]
Thank you. I've changed the configuration so it is no longer a volume and updated the PR.

File: library/eclipse-mosquitto [hunk 1]
```diff
@@ -1,7 +1,7 @@
-Maintainers: David Audet <david.audet@ca.com> (@daudetCA)
+Maintainers: Roger Light <roger@atchoo.org> (@ralight)
 GitRepo: https://github.com/eclipse/mosquitto.git
+GitCommit: 26e57661826f7dd3d965a5f8fea928d785a68b5b
+Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
```
Comment (yosifkit) [inline library/eclipse-mosquitto]: Doh, almost missed this one, the base image `alpine:3.8` doesn't support `arm32v7`.

Comment (ralight) [reply library/eclipse-mosquitto]: Well spotted! I've removed it and pushed the change.

**Comment by yosifkit** [comment]
LGTM cc @tianon

<details>
<summary>Diff:</summary>

```diff
diff --git a/_bashbrew-arches b/_bashbrew-arches
index d69cbdb..bd134e6 100644
--- a/_bashbrew-arches
+++ b/_bashbrew-arches
@@ -1 +1,6 @@
 eclipse-mosquitto:latest @ amd64
+eclipse-mosquitto:latest @ arm32v6
+eclipse-mosquitto:latest @ arm64v8
+eclipse-mosquitto:latest @ i386
+eclipse-mosquitto:latest @ ppc64le
+eclipse-mosquitto:latest @ s390x
diff --git a/_bashbrew-list b/_bashbrew-list
index 82c7fd6..55dfd82 100644
--- a/_bashbrew-list
+++ b/_bashbrew-list
@@ -1,2 +1,3 @@
-eclipse-mosquitto:1.4.12
+eclipse-mosquitto:1.5
+eclipse-mosquitto:1.5.3
 eclipse-mosquitto:latest
diff --git a/eclipse-mosquitto_latest/Dockerfile b/eclipse-mosquitto_latest/Dockerfile
index be20a14..6efa556 100644
--- a/eclipse-mosquitto_latest/Dockerfile
+++ b/eclipse-mosquitto_latest/Dockerfile
@@ -1,13 +1,69 @@
-FROM alpine:3.6
-MAINTAINER David Audet <david.audet@ca.com>
+FROM alpine:3.8
 
-LABEL Description="Eclipse Mosquitto MQTT Broker"
+LABEL maintainer="Roger Light <roger@atchoo.org>" \
+	 description="Eclipse Mosquitto MQTT Broker"
 
-RUN apk --no-cache add mosquitto=1.4.12-r0 && \
+ENV VERSION=1.5.3 \
+	 DOWNLOAD_SHA256=3081a998d303a883b1cd064009beabc88aa9159e26f5258a4ae6007160491d10 \
+	 GPG_KEYS=A0D6EEA1DCAE49A635A3B2F0779B22DFB3E717B7
+
+RUN set -x && \
+	 apk --no-cache add --virtual build-deps \
+      build-base \
+      curl \
+	  gnupg \
+      libressl-dev \
+      libwebsockets-dev \
+      util-linux-dev && \
+    curl -fSL https://mosquitto.org/files/source/mosquitto-${VERSION}.tar.gz -o /tmp/mosq.tar.gz && \
+	echo "$DOWNLOAD_SHA256  /tmp/mosq.tar.gz" | sha256sum -c - && \
+    curl -fSL https://mosquitto.org/files/source/mosquitto-${VERSION}.tar.gz.asc -o /tmp/mosq.tar.gz.asc && \
+	export GNUPGHOME="$(mktemp -d)" && \
+	found=''; \
+	for server in \
+		ha.pool.sks-keyservers.net \
+		hkp://keyserver.ubuntu.com:80 \
+		hkp://p80.pool.sks-keyservers.net:80 \
+		pgp.mit.edu \
+	; do \
+		echo "Fetching GPG key $GPG_KEYS from $server"; \
+		gpg --keyserver "$server" --keyserver-options timeout=10 --recv-keys "$GPG_KEYS" && found=yes && break; \
+	done; \
+	test -z "$found" && echo >&2 "error: failed to fetch GPG key $GPG_KEYS" && exit 1; \
+	gpg --batch --verify /tmp/mosq.tar.gz.asc /tmp/mosq.tar.gz && \
+	gpgconf --kill all && \
+	rm -rf "$GNUPGHOME" /tmp/mosq.tar.gz.asc && \
+	mkdir -p /build && \
+	tar --strip=1 -xf /tmp/mosq.tar.gz -C /build && \
+	rm /tmp/mosq.tar.gz && \
+    make -C /build -j "$(nproc)" \
+      WITH_ADNS=no \
+	  WITH_DOCS=no \
+      WITH_MEMORY_TRACKING=no \
+	  WITH_SHARED_LIBRARIES=no \
+      WITH_SRV=no \
+	  WITH_STRIP=yes \
+      WITH_TLS_PSK=no \
+      WITH_WEBSOCKETS=yes \
+      prefix=/usr \
+      binary && \
+	addgroup -S mosquitto 2>/dev/null && \
+    adduser -S -D -H -h /var/empty -s /sbin/nologin -G mosquitto -g mosquitto mosquitto 2>/dev/null && \
     mkdir -p /mosquitto/config /mosquitto/data /mosquitto/log && \
-    cp /etc/mosquitto/mosquitto.conf /mosquitto/config && \
-    chown -R mosquitto:mosquitto /mosquitto
+	install -d /usr/sbin/ && \
+	install -s -m755 /build/src/mosquitto /usr/sbin/mosquitto && \
+	install -m644 /build/mosquitto.conf /mosquitto/config/mosquitto.conf && \
+    sed -i -e 's/#log_dest stderr/log_dest syslog/' /mosquitto/config/mosquitto.conf && \
+    chown -R mosquitto:mosquitto /mosquitto && \
+	apk del build-deps && \
+	apk --no-cache add \
+      libuuid \
+      libwebsockets && \
+	rm -rf /build /etc/apk /lib/apk
+
+VOLUME ["/mosquitto/data", "/mosquitto/log"]
 
+# Set up the entry point script and default command
 COPY docker-entrypoint.sh /
 ENTRYPOINT ["/docker-entrypoint.sh"]
 CMD ["/usr/sbin/mosquitto", "-c", "/mosquitto/config/mosquitto.conf"]
diff --git a/eclipse-mosquitto_latest/docker-entrypoint.sh b/eclipse-mosquitto_latest/docker-entrypoint.sh
index 1a9fc8d..b381ac5 100755
--- a/eclipse-mosquitto_latest/docker-entrypoint.sh
+++ b/eclipse-mosquitto_latest/docker-entrypoint.sh
@@ -1,5 +1,4 @@
 #!/bin/ash
-
 set -e
-exec "$@"
 
+exec "$@"
```

</details>

Build test of #4987; 78dcb0c5; `amd64` (`eclipse-mosquitto`):

```console

$ bashbrew build eclipse-mosquitto:1.5.3
Building bashbrew/cache:2c61be152d0467ed3df38b6f44c90f3ed63b10fc54a4bf2696500b500c65fdc0 (eclipse-mosquitto:1.5.3)
Tagging eclipse-mosquitto:1.5.3
Tagging eclipse-mosquitto:1.5
Tagging eclipse-mosquitto:latest

$ test/run.sh eclipse-mosquitto:1.5.3
testing eclipse-mosquitto:1.5.3
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...passed
	'no-hard-coded-passwords' [3/4]...passed
	'override-cmd' [4/4]...passed

```

**Comment by tianon** [comment]
> ```diff
> +    sed -i -e 's/#log_dest stderr/log_dest syslog/' /mosquitto/config/mosquitto.conf && \
> ```

This seems kind of odd -- why are you switching from stderr to syslog?  Isn't stderr or stdout exactly what we want in a Docker environment?

> ```diff
> +	apk del build-deps && \
> +	apk --no-cache add \
> +      libuuid \
> +      libwebsockets && \
> ```

This is also a bit odd -- these two libraries are already installed before the `apk del`, and if you moved up the `apk add` to before the `apk del`, they wouldn't re-download (they would simply not be removed), although this is purely an optimization, it'll work as-is just fine (especially since `apk add` is really fast).

As a general "Alpine Linux" note, you can probably also drop `curl` and use the `wget` that comes pre-installed, but that doesn't really matter (just a note in case that's interesting to you).

**Comment by tianon** [comment]
<details>
<summary>Diff:</summary>

```diff
diff --git a/_bashbrew-arches b/_bashbrew-arches
index d69cbdb..bd134e6 100644
--- a/_bashbrew-arches
+++ b/_bashbrew-arches
@@ -1 +1,6 @@
 eclipse-mosquitto:latest @ amd64
+eclipse-mosquitto:latest @ arm32v6
+eclipse-mosquitto:latest @ arm64v8
+eclipse-mosquitto:latest @ i386
+eclipse-mosquitto:latest @ ppc64le
+eclipse-mosquitto:latest @ s390x
diff --git a/_bashbrew-list b/_bashbrew-list
index 82c7fd6..55dfd82 100644
--- a/_bashbrew-list
+++ b/_bashbrew-list
@@ -1,2 +1,3 @@
-eclipse-mosquitto:1.4.12
+eclipse-mosquitto:1.5
+eclipse-mosquitto:1.5.3
 eclipse-mosquitto:latest
diff --git a/eclipse-mosquitto_latest/Dockerfile b/eclipse-mosquitto_latest/Dockerfile
index be20a14..ef5418c 100644
--- a/eclipse-mosquitto_latest/Dockerfile
+++ b/eclipse-mosquitto_latest/Dockerfile
@@ -1,13 +1,67 @@
-FROM alpine:3.6
-MAINTAINER David Audet <david.audet@ca.com>
+FROM alpine:3.8
 
-LABEL Description="Eclipse Mosquitto MQTT Broker"
+LABEL maintainer="Roger Light <roger@atchoo.org>" \
+	 description="Eclipse Mosquitto MQTT Broker"
 
-RUN apk --no-cache add mosquitto=1.4.12-r0 && \
+ENV VERSION=1.5.3 \
+	 DOWNLOAD_SHA256=3081a998d303a883b1cd064009beabc88aa9159e26f5258a4ae6007160491d10 \
+	 GPG_KEYS=A0D6EEA1DCAE49A635A3B2F0779B22DFB3E717B7
+
+RUN set -x && \
+	 apk --no-cache add --virtual build-deps \
+      build-base \
+	  gnupg \
+      libressl-dev \
+      libwebsockets-dev \
+      util-linux-dev && \
+    wget https://mosquitto.org/files/source/mosquitto-${VERSION}.tar.gz -O /tmp/mosq.tar.gz && \
+	echo "$DOWNLOAD_SHA256  /tmp/mosq.tar.gz" | sha256sum -c - && \
+    wget https://mosquitto.org/files/source/mosquitto-${VERSION}.tar.gz.asc -O /tmp/mosq.tar.gz.asc && \
+	export GNUPGHOME="$(mktemp -d)" && \
+	found=''; \
+	for server in \
+		ha.pool.sks-keyservers.net \
+		hkp://keyserver.ubuntu.com:80 \
+		hkp://p80.pool.sks-keyservers.net:80 \
+		pgp.mit.edu \
+	; do \
+		echo "Fetching GPG key $GPG_KEYS from $server"; \
+		gpg --keyserver "$server" --keyserver-options timeout=10 --recv-keys "$GPG_KEYS" && found=yes && break; \
+	done; \
+	test -z "$found" && echo >&2 "error: failed to fetch GPG key $GPG_KEYS" && exit 1; \
+	gpg --batch --verify /tmp/mosq.tar.gz.asc /tmp/mosq.tar.gz && \
+	gpgconf --kill all && \
+	rm -rf "$GNUPGHOME" /tmp/mosq.tar.gz.asc && \
+	mkdir -p /build && \
+	tar --strip=1 -xf /tmp/mosq.tar.gz -C /build && \
+	rm /tmp/mosq.tar.gz && \
+    make -C /build -j "$(nproc)" \
+      WITH_ADNS=no \
+	  WITH_DOCS=no \
+      WITH_MEMORY_TRACKING=no \
+	  WITH_SHARED_LIBRARIES=no \
+      WITH_SRV=no \
+	  WITH_STRIP=yes \
+      WITH_TLS_PSK=no \
+      WITH_WEBSOCKETS=yes \
+      prefix=/usr \
+      binary && \
+	addgroup -S mosquitto 2>/dev/null && \
+    adduser -S -D -H -h /var/empty -s /sbin/nologin -G mosquitto -g mosquitto mosquitto 2>/dev/null && \
     mkdir -p /mosquitto/config /mosquitto/data /mosquitto/log && \
-    cp /etc/mosquitto/mosquitto.conf /mosquitto/config && \
-    chown -R mosquitto:mosquitto /mosquitto
+	install -d /usr/sbin/ && \
+	install -s -m755 /build/src/mosquitto /usr/sbin/mosquitto && \
+	install -m644 /build/mosquitto.conf /mosquitto/config/mosquitto.conf && \
+    chown -R mosquitto:mosquitto /mosquitto && \
+	apk --no-cache add \
+      libuuid \
+      libwebsockets && \
+	apk del build-deps && \
+	rm -rf /build /etc/apk /lib/apk
+
+VOLUME ["/mosquitto/data", "/mosquitto/log"]
 
+# Set up the entry point script and default command
 COPY docker-entrypoint.sh /
 ENTRYPOINT ["/docker-entrypoint.sh"]
 CMD ["/usr/sbin/mosquitto", "-c", "/mosquitto/config/mosquitto.conf"]
diff --git a/eclipse-mosquitto_latest/docker-entrypoint.sh b/eclipse-mosquitto_latest/docker-entrypoint.sh
index 1a9fc8d..b381ac5 100755
--- a/eclipse-mosquitto_latest/docker-entrypoint.sh
+++ b/eclipse-mosquitto_latest/docker-entrypoint.sh
@@ -1,5 +1,4 @@
 #!/bin/ash
-
 set -e
-exec "$@"
 
+exec "$@"
````

</details>

**Comment by tianon** [comment]
LGTM :tada: :balloon: 

Build test of #4987; a98c5d22; `amd64` (`eclipse-mosquitto`):

```console

$ bashbrew build eclipse-mosquitto:1.5.3
Building bashbrew/cache:280abc42d52ed722763531ed447d3cf2c237b15126db1ad7086a4308ae79090d (eclipse-mosquitto:1.5.3)
Tagging eclipse-mosquitto:1.5.3
Tagging eclipse-mosquitto:1.5
Tagging eclipse-mosquitto:latest

$ test/run.sh eclipse-mosquitto:1.5.3
testing eclipse-mosquitto:1.5.3
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...passed
	'no-hard-coded-passwords' [3/4]...passed
	'override-cmd' [4/4]...passed

```

**Pull request merged**